### PR TITLE
rgw: add initial rgw wrappers and tests

### PR DESCRIPTION
**Pull request by yanghonggang**
Signed-off-by: Yang Honggang <yanghonggang@kuaishou.com>

## Checklist
- [ ] Added tests for features and functional changes
- [ ] Public functions and types are documented
- [ ] Standard formatting is applied to Go code

### CODE REVIEW

**Comment by mattbenjamin** [comment]
hi @yanghonggang ; so, this is a go binding to librgw?  maybe commit msg/pr could highlight that--took me a minute to see what this actually was

**Comment by mattbenjamin** [comment]
(sorry, wrong repo :)

**Comment by yanghonggang** [comment]
> (sorry, wrong repo :)

I want to add librgw go binding into go-ceph. 
`make check` failed,  I will try to fix these errors...

**Comment by phlogistonjohn** [comment]
Hi @yanghonggang, I'm very excited to see this! It wasn't until recently that I was even aware there was a librgw (tucked away in the rados subdir), so I'm enthused to see this come from a new contributor.

That said, I'd like to set expectations: this is a pretty large new chunk of code and we are rapidly coming up on the Christmas & New Years holidays. This is likely going to need thorough review. Many of us who maintain go-ceph do so as part of $dayjob and are taking time off from work. So between these points I caution you to expect this to take some time. Please don't let that discourage you though. I would love to see this be part of our next release as a marquee feature. Please be patient with us. :-)


Now, before we get around to formally reviewing the patches line-by-line here are a few general tips:
* Get the CI jobs to pass:
  * Focus on "check" first, it is pretty easy and fast to run [revive](https://github.com/mgechev/revive) locally with our makefile rule (make check) 
  * Once check is passing, work on getting the test suites to pass if the above doesn't automatically resolve the issues
* Use the checklist in the PR template to help you ensure documentation and other rules are applied. While revive helps enforce much of this now, it can still help to check items off the checklist once you've double checked them.
* It can make review go faster if the commits are smaller. To avoid burdening you a lot I suggest a simple rule to organize by files: one commit for the library code, one commit for the tests, one commit for the "infrastructure" (the makefile, scripts, etc) that glue the code and tests into the CI.
* Please provide a PR description. It doesn't have to be long, but it should briefly describe what makes up the PR and why you want to add it. We can certainly tell a lot from the patches and commit messages, but a introduction can help all reviewers get up to speed, esp. if you provide reference links.

Somewhat coincidentally, I am aware we're lacking in a strong contributor's guide and that was a task I plan to take on in the new year. Feel free to ask questions about the process or the structure in go-ceph if anything seems unclear. Anything you ask may help us make a better guide for future contributors!

Thanks, and we'll try to get back to you soon.

**Comment by yanghonggang** [comment]
> Hi @yanghonggang, I'm very excited to see this! It wasn't until recently that I was even aware there was a librgw (tucked away in the rados subdir), so I'm enthused to see this come from a new contributor.
> 
> That said, I'd like to set expectations: this is a pretty large new chunk of code and we are rapidly coming up on the Christmas & New Years holidays. This is likely going to need thorough review. Many of us who maintain go-ceph do so as part of $dayjob and are taking time off from work. So between these points I caution you to expect this to take some time. Please don't let that discourage you though. I would love to see this be part of our next release as a marquee feature. Please be patient with us. :-)
> 
> Now, before we get around to formally reviewing the patches line-by-line here are a few general tips:
> 
> * Get the CI jobs to pass:
>   
>   * Focus on "check" first, it is pretty easy and fast to run [revive](https://github.com/mgechev/revive) locally with our makefile rule (make check)
>   * Once check is passing, work on getting the test suites to pass if the above doesn't automatically resolve the issues
> * Use the checklist in the PR template to help you ensure documentation and other rules are applied. While revive helps enforce much of this now, it can still help to check items off the checklist once you've double checked them.
> * It can make review go faster if the commits are smaller. To avoid burdening you a lot I suggest a simple rule to organize by files: one commit for the library code, one commit for the tests, one commit for the "infrastructure" (the makefile, scripts, etc) that glue the code and tests into the CI.
> * Please provide a PR description. It doesn't have to be long, but it should briefly describe what makes up the PR and why you want to add it. We can certainly tell a lot from the patches and commit messages, but a introduction can help all reviewers get up to speed, esp. if you provide reference links.
> 
> Somewhat coincidentally, I am aware we're lacking in a strong contributor's guide and that was a task I plan to take on in the new year. Feel free to ask questions about the process or the structure in go-ceph if anything seems unclear. Anything you ask may help us make a better guide for future contributors!
> 
> Thanks, and we'll try to get back to you soon.

OK, thank you

**Comment by phlogistonjohn** [comment]
In case I was too wordy with my last reply and unclear: Please do resolve the CI failures starting with the "check" rule. This should be done before we start looking at the PR in depth. Thanks!

**Comment by yanghonggang** [comment]
> In case I was too wordy with my last reply and unclear: Please do resolve the CI failures starting with the "check" rule. This should be done before we start looking at the PR in depth. Thanks!
OK

**Comment by yanghonggang** [comment]
> In case I was too wordy with my last reply and unclear: Please do resolve the CI failures starting with the "check" rule. This should be done before we start looking at the PR in depth. Thanks!

@phlogistonjohn 
Done

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:1]: I strongly recommend splitting this file up into multiple smaller files. It makes the code simpler to understand and clearer what C functions are used by what.

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:21]: We have our own internal callback helper library under `github.com/ceph/go-ceph/internal/callbacks`. Examples can be found in `rbd/diff_iterate.go` and `rbd/watchers_mimic.go`. If you have any other questions about that callback helper library please ask, but it is requested that you use that rather than rely on a different external dependency.

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:54]: IIUC this will pass go memory pointing at C memory to a C function. This seems like it may work but will fail the stricter cgo pointer passing rules when run with GODEBUG=cgocheck=2 (see PR #411).  Until that PR is merged, we should take care to make sure we don't add new violations. Please run the tests with that environment variable when possible.

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:61]: To match the other packages in go-ceph, I think this should be a method on the RGW type.

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:76]: The word "keep" here is a bit confusing. Perhaps "MountFlagNone selects the default behavior of Mount" would be better?

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:106]: Again, "keep" is confusing. See above please.

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:500]: Methods with common names to Go interfaces should probably match the common interfaces. For example io.Writer defines Write as `Write(p []byte) (n int, err error)`. On top of that, we want the various sub-packages of go-ceph to be reasonably consistent. Therefore I think that this would be better named as WriteAt and exist as a method on the FileHandle type.

File: rgw/rgw.go
Comment (phlogistonjohn) [inline rgw/rgw.go:529]: See my comment on Write.

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
Hi there! I've added a few initial comments on the PR. First off, I want to say that there looks like a lot of good stuff here.
It's a big PR so I didn't comment on everything I think it needs yet, but more of the very high level items that can be improved first.

Please do ask questions or comment back on any items you think are unclear or wish do discuss further!

**Comment by yanghonggang** [comment]
> Hi there! I've added a few initial comments on the PR. First off, I want to say that there looks like a lot of good stuff here.
> It's a big PR so I didn't comment on everything I think it needs yet, but more of the very high level items that can be improved first.
> 
> Please do ask questions or comment back on any items you think are unclear or wish do discuss further!

Thank you for the comments. I will try to fix all these problems.

**Comment by phlogistonjohn** [comment]
Hi @yanghonggang,
Since you haven't updated this PR in a while I was wondering if you'd object to the maintainers (mainly me and possibly a few others) taking small parts of the PR as new separate PRs? I ask because a) it would really be nice to have some basic RGW support in go ceph b) I have been asked about some rgw management apis and having rgw data apis would be good to test those, etc. c) just for the sake of completeness and I would like to see your work bear fruit.

I would want to make sure you are credited in the commits too, of course.

Any thoughts?

**Comment by yanghonggang** [comment]
> Hi @yanghonggang,
> Since you haven't updated this PR in a while I was wondering if you'd object to the maintainers (mainly me and possibly a few others) taking small parts of the PR as new separate PRs? I ask because a) it would really be nice to have some basic RGW support in go ceph b) I have been asked about some rgw management apis and having rgw data apis would be good to test those, etc. c) just for the sake of completeness and I would like to see your work bear fruit.
> 
> I would want to make sure you are credited in the commits too, of course.
> 
> Any thoughts?

Sorry for the late response. I'm busy finding a new job these days :). Feel free to do this. You can do whatever to this PR.

**Comment by phlogistonjohn** [comment]
> > Hi @yanghonggang,
> > Since you haven't updated this PR in a while I was wondering if you'd object to the maintainers (mainly me and possibly a few others) taking small parts of the PR as new separate PRs? I ask because a) it would really be nice to have some basic RGW support in go ceph b) I have been asked about some rgw management apis and having rgw data apis would be good to test those, etc. c) just for the sake of completeness and I would like to see your work bear fruit.
> > I would want to make sure you are credited in the commits too, of course.
> > Any thoughts?
> 
> @phlogistonjohn
> Sorry for the late response. I'm busy finding a new job these days :). Feel free to do this. You can do whatever to this PR.

Understood, thanks for the update. I do appreciate your efforts!
I will discuss the PR itself with the rest of the maintenance team soon.

**Comment by phlogistonjohn** [comment]
As per the earlier discussion this PR will be closed.
Until additional requirements or advantages of wrapping the RGW C layer is found we recommend the use of S3 compatible API of RGW with an S3 compatible library in Go for accessing data using RGW.

**Pull request closed without merging**
